bento/src/main/java/com/yelp/android/bento/core/AsyncInflationBridge.kt

```diff
@@ -0,0 +1,106 @@
+ * is finished.
+ */
+internal class AsyncInflationBridge(val recyclerView: RecyclerView) {
+
+    private val context = recyclerView.context
+    private var coroutineScope: CoroutineScope = if (context is LifecycleOwner) {
+        context.lifecycleScope
+    } else {
+        GlobalScope
+    }
+
+    private companion object {
+        private val lock = Mutex()
+        private val dispatcher = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+    }
+
+    private val viewHolderMap = mutableMapOf<Class<out ComponentViewHolder<*, *>>,
+            MutableList<ComponentViewHolder<*, *>>>()
+    private val viewMap = mutableMapOf<ComponentViewHolder<*, *>, View?>()
+
+    /**
+     * Inflates the views and creates the view holder for [component] on a background thread,
+     * getting them nice and ready for RecyclerViewComponentController to use them.
+     */
+    fun asyncInflateViewsForComponent(component: Component, addComponentCallback: () -> Unit) {
+        coroutineScope.launch(dispatcher) {
+            val inflations = (0 until component.count).map { i ->
+                async {
+                    val viewHolderType = component.getHolderType(i)
+                    val viewHolder: ComponentViewHolder<*, *> = constructViewHolder(viewHolderType)
+                    addViewHolder(viewHolder, viewHolderType)
+                    val (_, view) = BentoAsyncLayoutInflater.inflate(viewHolder, recyclerView)
+                    viewMap[viewHolder] = view
+                }
+            }
+            lock.withLock {
+                inflations.awaitAll()
+            }
+            withContext(Dispatchers.Main) {
+                try {
+                    addComponentCallback()
+                } catch (e: IllegalArgumentException) {
+                    // TODO: handle crash if component is already added.
+                }
+            }
+        }
+    }
+
+    /**
+     * Retrieves a previously async inflated view if there is one.
+     */
+    @Synchronized
+    fun getView(viewHolder: ComponentViewHolder<*, *>) = viewMap[viewHolder]
+
+    /**
+     * Retrieves a previously created view holder of the passed in [viewHolderType].
+     */
+    fun getViewHolder(
+            viewHolderType: Any? // Class<out ComponentViewHolder<*, *>> doesn't work with Java.
+    ): ComponentViewHolder<*, *>? {
+        if (viewHolderMap.containsKey(viewHolderType)) {
+            viewHolderMap[viewHolderType]?.let { list ->
+                if (list.isNotEmpty()) {
+                    return list.removeAt(0)
+                }
+            }
+        }
+        return null
+    }
+
+    private fun addViewHolder(
+            viewHolder: ComponentViewHolder<*, *>,
+            viewHolderType: Class<out ComponentViewHolder<*, *>>
+    ) {
+        var viewHolders = viewHolderMap[viewHolderType]
+        if (viewHolders == null) {
+            viewHolders = mutableListOf()
+        }
```

Comment:
More kotliny:

```
val viewHolders = viewHolderMap.getOrPut(viewHolderType){ mutableListOf() }
```
And as we said through slack, a concurrent list would probably be best

Reply:
Or at last, synchronize read/write operation in the list.

Reply:
So maybe this is sufficient: 

```
fun getViewHolder(
    viewHolderType: Any? // Class<out ComponentViewHolder<*, *>> doesn't work with Java.
): ComponentViewHolder<*, *>? {
    if (viewHolderMap.containsKey(viewHolderType)) {
        viewHolderMap[viewHolderType]?.let { list ->
            if (list.isNotEmpty()) {
                synchronized(this) {
                    return list.removeAt(0)
                }
            }
        }
    }
    return null
}

private fun addViewHolder(
    viewHolder: ComponentViewHolder<*, *>,
    viewHolderType: Class<out ComponentViewHolder<*, *>>
) {
    val viewHolders = viewHolderMap.getOrPut(viewHolderType) {
        mutableListOf()
    }
    synchronized(this) {
        viewHolders.add(viewHolder)
    }
    viewHolderMap[viewHolderType] = viewHolders
}
```

Reply:
I'm using a `ConcurrentLinkedDeque` for the list. I'll update it to use `getOrPut` too 👍